spec.bs

```diff
@@ -1,16 +1,255 @@
+
+Modify the [=determine the storage access policy=] step to insert the following steps before step 3 (running [=implementation-defined=] steps):
+
+1. Let |embeddedSite| be the result of [=obtain a site|obtaining a site=] from key’s embedded origin.
+2. Let |sameSet| be true if |embeddedSite| is [=eligible for same-party membership when embedded within=] |topLevelSite|.
+3. Optionally set implicitly granted or implicitly denied based on the value of |sameSet|. This step is [=implementation-defined=].
+
+A [=site=] |embeddedSite| is <dfn export>eligible for same-party membership when embedded within</dfn> a [=site=] |topLevelSite|, if the following steps return true:
+
+1. Let |set| be the result of [=find a first-party set|finding a first-party set=] for |topLevelSite|.
+2. If |set| is null, return false.
+3. Let |topLevelType| be the [=determine the member type|member type=] of |topLevelSite| in |set|.
+4. If |topLevelType| is “associated” and the result of [=determine eligibility for an associated site|determining eligibility for an associated site=] given |topLevelSite| and |set| is false, return false.
+5. If |topLevelType| is “service”, return false.
+6. Let |type| be the [=determine the member type|member type=] of |embeddedSite| in |set|.
+7. If |type| is “none”, return false.
+8. If |type| is “associated”, return the result of [=determine eligibility for an associated site|determining eligibility for an associated site=] given |embeddedSite| and |set|.
+9. Return true.
+
+To <dfn>determine eligibility for an associated site</dfn> given a [=site=] |site| and a [=first-party set=] |set|, run the following steps:
+
+1. If |set|’s [=first-party set/associatedSites=] does not contain |site|, return false.
+2. Let |index| be the index of |site| in |set|’s [=first-party set/associatedSites=].
+3. If |index| is greater than or equal to the [=limit for associated sites=], return false.
+4. Return true.
+
+<h2 id="handling-changes">Handling first-party set changes</h2>
+
+When a [=site=] |site| leaves a [=first-party set=] as the result of building a new [=list of first-party sets=], user agents must ensure that it does not retain any access to data or shared identifiers held by other sites in the first-party set by running the following steps:
+
+1. Assert that |site| is not an [=opaque origin=].
+2. Let |domain| be site’s [=origin/host=].
+3. For each |origin| known to the user agent whose [=origin/host=]'s [=registered domain=] is |domain|:
+    1. [=Clear cache for origin=].
+    2. [=Clear cookies for origin=].
+    3. [=Clear DOM-accessible storage for origin=].
+    4. Let |descriptor| be a newly-created {{PermissionDescriptor}} with {{PermissionDescriptor/name}} initialized to “storage-access”.
+    5. Remove all permission store entries for |descriptor|, where key[0] is |site| or key[1] is |origin|.
```

Comment:
I'm not sure how closely aligned this needs to be with practical implementation details; but I believe with recent spec discussions on SAA moving to a per-frame model, the storage-access permission doesn't persist across browser restarts (except where browser features such as session-restores occur, but I'm not sure if that feature is specified), and so this type of clearing may not be required?

cc @shuranhuang

Reply:
Browsers do persist the storage-access permission across restarts; IIRC Safari's permission lasts 30 days of usage time. (Maybe session-restore gets involved there such that it's actually considered a single 30-day-long session, but I don't think we can assume that.)

However, as long as the browser's behavior is indistinguishable from what it would be if the browser implemented the spec exactly, it's fine. So a browser is free to implement this part of the spec as "delete the matching permission store entries", or as "don't bother to persist these permission store entries", since they both give the same observable behavior.

Reply:
I agree with what Chris said. The permissions should be listed out as a clearing item in the spec, so that the actual implementation from different browsers does not matter as long as the behavior is indistinguishable. To add more implementation detail, right now Chrome and Edge both persist the SAA (implicit) permissions, but when a new user session is created during browser restarts, Edge can continue the life of those permissions if the restore feature is enabled, whereas Chrome removes those permissions (autogranted by FPS) regardless the restore feature.

Reply:
Thanks @cfredric and @shuranhuang . Okay, I'm not sure what's recommended for specs in such situations; but I'll leave it to @johannhof 's judgement.

Reply:
I think it's important to list this even if some implementations may have different lifetimes on the permission, to avoid leakage :)